File: examples/python/pwa-flask/template.py
"""
    Template and response
    See: https://github.com/FranBarInstance/neutralts-docs
"""

import json

from constants import TEMPLATE_ERROR, TEMPLATE_ROUTER
from flask import make_response, redirect
from neutraltemplate import NeutralTemplate


class Template:
    """NeutralTemplate wrapper"""
    def __init__(self, schema):
        self.schema = schema

    def render(self):
        """render template and return response"""
        template = NeutralTemplate(TEMPLATE_ROUTER, json.dumps(self.schema))
        contents = template.render()
        status_code = int(template.get_status_code())
        status_text = template.get_status_text()
        status_param = template.get_status_param()

        # The template may generate redirects.
        if status_code in [301, 302, 307, 308]:
            response = make_response(redirect(status_param, code=status_code))
            return response

        # The template may generate HTTP errors.
        if status_code >= 400:
            error = {
                "data": {
                    "CONTEXT": {
                        "ROUTE": "error"
                    },
                    "error": {
                        "code": status_code,
                        "text": status_text,
                        "param": status_param
                    }
                }
            }
            template.set_path(TEMPLATE_ERROR)
            template.merge_schema(json.dumps(error))
            contents = template.render()

        response = make_response(contents, status_code)
        response.headers['Content-Type'] = 'text/html'
        return response

File: examples/rust/pwa-warp/src/template.rs
//! Template and response
//! See: https://github.com/FranBarInstance/neutralts-docs

use warp::http::StatusCode;
use warp::http::Uri;
use neutralts::Template as NeutralTemplate;
use serde_json::{json, Value};
use crate::constants::{TEMPLATE_ERROR, TEMPLATE_ROUTER};

pub struct Template {
    schema: Value,
}

impl Template {
    pub fn new(schema: Value) -> Self {
        Self { schema }
    }

    pub fn render(&self) -> Result<impl warp::Reply, Box<dyn std::error::Error>> {

        // Create template from file with schema
        let mut template = NeutralTemplate::from_file_value(TEMPLATE_ROUTER, self.schema.clone())?;
        let contents = template.render();

        // Get the status code.
        // Neutral TS may send an HTTP error code, with {:redirect; ... :} or {:exit; ... :}
        let status_code = template.get_status_code();
        let status_text = template.get_status_text();
        let status_param = template.get_status_param();

        if let Ok(code) = status_code.parse::<u16>() {

            // The template may generate redirects.
            if [301, 302, 307, 308].contains(&code) {
                let uri: Uri = status_param.parse().unwrap();
                let _response = warp::redirect::temporary(uri);
                let response = warp::reply::html(format!("<html><body>Redirecting to {}</body></html>", status_param));
                return Ok(warp::reply::with_status(response, StatusCode::from_u16(code).unwrap()));
            }

            // The template may generate HTTP errors.
            if code >= 400 {
                let error = json!({
                    "data": {
                        "CONTEXT": {
                            "ROUTE": "error"
                        },
                        "error": {
                            "code": code,
                            "text": status_text,
                            "param": status_param
                        }
                    }
                });

                // Again with error page
                template.set_src_path(TEMPLATE_ERROR).unwrap();

                // Error variables are added to the schema.
                template.merge_schema_value(error);

                // Render the error page
                let error_contents = template.render();

                let response = warp::reply::html(error_contents);
                return Ok(warp::reply::with_status(response, StatusCode::from_u16(code).unwrap()));
            }
        }

        // 200 Ok
        let response = warp::reply::html(contents);
        Ok(warp::reply::with_status(response, StatusCode::OK))
    }
}
